File: src/services/audit/detectors/ad/accounts/advanced/privileged-account-spn.ts
/**
 * Privileged Account SPN Detector
 * Detect privileged accounts with SPNs (kerberoastable)
 * Admin accounts should NOT have SPNs as they become kerberoasting targets
 */

import { ADUser } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedUserEntities } from '../../../../../../utils/entity-converter';

export function detectPrivilegedAccountSpn(users: ADUser[], includeDetails: boolean): Finding {
  const affected = users.filter((u) => {
    // Must be privileged (adminCount=1)
    if (u.adminCount !== 1) return false;
    // Must be enabled
    if (!u.enabled) return false;

    // Must have SPN
    const spn = u['servicePrincipalName'];
    const hasSPN = spn && Array.isArray(spn) && spn.length > 0;

    return hasSPN;
  });

  return {
    type: 'PRIVILEGED_ACCOUNT_SPN',
    severity: 'high',
    category: 'accounts',
    title: 'Privileged Account with SPN',
    description:
      'Privileged accounts (adminCount=1) have Service Principal Names configured. These accounts are vulnerable to Kerberoasting attacks.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedUserEntities(affected) : undefined,
    details:
      affected.length > 0
        ? {
            attackVector: 'Request TGS ticket → Offline crack password → Full admin access',
            recommendation:
              'Remove SPNs from admin accounts. Use dedicated service accounts (preferably gMSA) for services.',
            criticalRisk: 'Compromising these accounts grants immediate Domain Admin or equivalent access.',
          }
        : undefined,
  };
}

File: internal/audit/detectors/ad/accounts/advanced/privileged-account-spn.go
package advanced

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// PrivilegedAccountSpnDetector detects privileged accounts with SPNs
type PrivilegedAccountSpnDetector struct {
	audit.BaseDetector
}

// NewPrivilegedAccountSpnDetector creates a new detector
func NewPrivilegedAccountSpnDetector() *PrivilegedAccountSpnDetector {
	return &PrivilegedAccountSpnDetector{
		BaseDetector: audit.NewBaseDetector("PRIVILEGED_ACCOUNT_SPN", audit.CategoryAccounts),
	}
}

// Detect executes the detection
func (d *PrivilegedAccountSpnDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.User

	for _, u := range data.Users {
		// Must be privileged (adminCount=true)
		if !u.AdminCount {
			continue
		}
		// Must be enabled
		if u.Disabled {
			continue
		}

		// Must have SPN
		if len(u.ServicePrincipalNames) > 0 {
			affected = append(affected, u)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "Privileged Account with SPN",
		Description: "Privileged accounts (adminCount=1) have Service Principal Names configured. These accounts are vulnerable to Kerberoasting attacks.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedUserEntities(affected)
		finding.Details = map[string]interface{}{
			"attackVector":    "Request TGS ticket -> Offline crack password -> Full admin access",
			"recommendation":  "Remove SPNs from admin accounts. Use dedicated service accounts (preferably gMSA) for services.",
			"criticalRisk":    "Compromising these accounts grants immediate Domain Admin or equivalent access.",
		}
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewPrivilegedAccountSpnDetector())
}
